### ISSUE TITLE
Race in close in the socket backend

### ISSUE THREAD

**Issue by samoht** [mirage/mirage-tcpip#347]
See https://github.com/mirage/ocaml-cohttp/pull/589

It seems that close can raise `Unix.EBADF`: we should catch it properly instead of propagating it to the rest of the stack.

**Comment by djs55** [comment]
Is the `EBADF` caused by closing a fd that's already closed? If so it could
be even worse -- the same fd number could be re-used by a concurrent call
to `open`, and we would close it by mistake.

We could wrap the `fd` in something like
```
type t = {
  mutable fd: Unix.file_descr option;
}
let close t =
  match t.fd with
  | None -> Log.warn (fun f -> f "log double-close?")
  | Some fd -> Unix.close fd; t.fd <- None
```


On Thu, Dec 7, 2017 at 12:17 PM, Thomas Gazagnaire <notifications@github.com

**Comment by ansiwen** [comment]
Related discussion:
https://discuss.ocaml.org/t/system-calls-unix-close-signals-eintr/972

**Comment by hannesm** [comment]
this is fixed by #366 -- unclear whether you guys would like to have a more appropriate fix (the suggested `close` is still racy, if `close` returns a unix error and does _not_ close the fd, the fd will leak)... but as mentioned in https://discuss.ocaml.org/t/system-calls-unix-close-signals-eintr/972 `close` on a unix is hard to do properly

**Comment by hannesm** [comment]
#366 is merged, closing this issue.
